### PR TITLE
Show a toast message (notice) when moving posts to trash

### PR DESCRIPTION
**Pull request by crazytonyli**
For context, this change is initiated from this conversation: p1694476848843439-slack-C0180B5PRJ4

## Changes

This PR changes how to present an "undo" action to users when moving a post to trash. Here is a video preview of before & after the change.

| Before (with an undo row)  | After (with a toast message) |
| ------------- | ------------- |
| <video src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1101828/7b781e26-cab4-4401-834b-f4561122fe43"></video>  | <video src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1101828/1df59f59-c018-4b86-b25f-488db2029931"></video>  |

### Problems of the "Undo" row

From design perspective, there are a couple of issues:
1. This feature here is the only one in the app that shows an "undo" row. In other actions in the posts list and other areas in the app, a toast message with an action button is commonly used.
2. After moving many posts to trash, the trashed posts' corresponding "undo" row has exactly the same appearance which is not intuitive to users when they do want to undo the previous trash action.

From implementation perspective, it's not straight-forward to implement keeping the trashed post position and placing a "undo" row in it. From a quick testing, the existing implementation has an obvious bug when trashing posts from bottom to up.

### New flow

As you can see in the video, the new design is showing a toast message with an "Undo" button, which is consistent with other actions like restoring post from trash.

### Implementation details

There are three parts involved in this PR

#### 1. Removing the `restorableStatus` stored property from `AbstractPost`

This property is used to kept the original post status when it's moved to trash. 

One obvious issue is, this property is not a "managed" property, which means it's not stored into the Core Data database. So, its value is guaranteed to be lost if multiple context objects are used to write and read this information.

Also, this property gives a false sense of a trashed post can be "restored" to _any_ status, but I believe in reality a trashed post can only be restored to draft.

#### 2. Use delete, trash, restore functions in `PostRepository`

These functions a newly implemented in `PostRepository`, as replacements for the ones in `PostService` which are now deleted in this PR.

#### 3. Show toast messages after trashing posts

This is the UI change part of this PR.

## Test Instructions

Here are a few things that I can think of:
1. Verify the "Undo" button can restore the trashed post as draft.
1. Verify that there is no side-effect in perform "Move to trash" action multiple times on the same post.
1. Verify trashing many posts at the same time works.

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
What's described in the "Test Instructions" section.

3. What automated tests I added (or what prevented me from doing so)
None.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/Classes/Services/PostRepository.swift
Comment (mokagio) [inline WordPress/Classes/Services/PostRepository.swift:86]: At first I was surprised to see the `assert` here and not at the start of the method. But I guess the reason is that the code above recur to find the original post, and for the same reason we want to delete the original, we want to check the status on it. Am I reading this right?

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift
Comment (mokagio) [inline WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift:985]: **Out of scope**: I've been thinking about all the accesses to `ContextManager.shared` that we have and the issues they sometime cause such as https://github.com/wordpress-mobile/WordPress-iOS/issues/21742.

Would it make sense to instantiate the view controller with a context or `CoreDataStack` object? In practice, this would be `ContextManager.shared.mainContext` or `ContextManager.shared` but the separation would:

1. Centralize the definition of the shared resource access
2. Allow us to replace it in the tests

What do you think?

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift
Comment (mokagio) [inline WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift]: This would be a good time to introduce reverse-DNS keys in these localized strings. For context: #19028

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift
Comment (mokagio) [inline WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift:1029]: Have you considered moving some of the logic here into a ViewModel-like object, so that we can test it?

One point of friction I can see is that there is no such object for this VC that I can see at the moment. Another is that this is a base class, meaning that the actual subclasses the app uses at runtime might need additional behavior and might be left with having to create their own ViewModel subclasses. This would be confusing. At the same time, it would be a sign of an abstraction that no longer serves us well and we might want to update its design.

**Review by mokagio — COMMENTED** [review]
Left a few questions.

File: WordPress/Classes/Services/PostRepository.swift
Comment (crazytonyli) [reply WordPress/Classes/Services/PostRepository.swift:86]: Yes. This status check doesn't exist in the original delete function in `PostService`. I added it here to catch incorrect calls to this function (which may never happen).

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift
Comment (crazytonyli) [reply WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift]: These strings were copied from https://github.com/wordpress-mobile/WordPress-iOS/pull/21724/files/aedccc4558780dbd690d4e2f586536f1b0ea8a97#diff-0b401d58526d69f6fa54a6679880b35a5fc7bf7643731e9c55f5d4b7b56f8d63L34-L35. I have updated them in https://github.com/wordpress-mobile/WordPress-iOS/pull/21724/commits/58c1d44c73dcb3709daaa8ade5d2e027b1ada652

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift
Comment (crazytonyli) [reply WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift:1029]: That'd be much better than this view controller's current state. But I don't want to just move this small part (trashing, deleting, restoring) out, leaving this view controller in a weird state with mixed-patterns. A large refactor on this screen also seems unnecessary at this point.

File: WordPress/Classes/Services/PostRepository.swift
Comment (mokagio) [reply WordPress/Classes/Services/PostRepository.swift:86]: 👍

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift
Comment (mokagio) [reply WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift]: Thank you

File: WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift
Comment (mokagio) [reply WordPress/Classes/ViewRelated/Post/AbstractPostListViewController.swift:1029]: > leaving this view controller in a weird state with mixed-patterns

Good point 🤔 

I am a fan of _sprouting_ (a la [Working Effectively with Legacy Code](https://www.amazon.com/Working-Effectively-Legacy-Michael-Feathers/dp/0131177052)) better code from existing code, even if what is added/extracted is a little portion of the total. But ☝️ having mixed patterns is a concern.

I guess it would be acceptable only if we had an agreed upon expectation that "all new and modified code should follow <pattern>." That is not the case at the moment.

**Review by mokagio — APPROVED** [review]
Given this PR has been sitting idle waiting for my feedback, the least I could do was fixing the merge conflict (on `RELEASE-NOTES.txt`) to unblock it's merge

**Pull request merged**
